convert-hf-to-gguf.py

```diff
@@ -2417,8 +2372,8 @@ def parse_args() -> argparse.Namespace:
         help="path to write to; default: based on input",
     )
     parser.add_argument(
-        "--outtype", type=str, choices=["f32", "f16"], default="f16",
-        help="output format - use f32 for float32, f16 for float16",
+        "--outtype", type=str, choices=["f32", "f16", "bf16"], default="f16",
```

Comment:
Given most models come in bf16 wouldn't it make sense to set it as default?

Reply:
The conversion to `bf16` is slightly slower and uses a bit more RAM than `f16` conversion, due to the lack of native Numpy support, so I didn't change the default.

I'll see if I can auto-detect whether the model contains `bf16` tensors (but it will most likely be too complicated). Otherwise, it does make sense to set `bf16` as default if it's very widely used.

Reply:
We only have rudimentary CPU-only `bf16` support in `ggml`, so `f16` is better for now

Reply:
when you say CPU-only, I assume you're referring to inference, since all conversion and quantization is currently CPU-only?

Reply:
> I'll see if I can auto-detect whether the model contains `bf16` tensors (but it will most likely be too complicated). Otherwise, it does make sense to set `bf16` as default if it's widely used.

@compilade could we not attempt to read from config.json? it should have a torch_dtype in it

Reply:
(left comment on wrong account)

>I'll see if I can auto-detect whether the model contains bf16 tensors (but it will most likely be too complicated). Otherwise, it does make sense to set bf16 as default if it's widely used.

@compilade could we not attempt to read from config.json? it should have a torch_dtype in it

Ggerganov, when you say CPU-only, I assume you're referring to inference, since all conversion and quantization is currently CPU-only?

Reply:
> could we not attempt to read from config.json? it should have a torch_dtype in it

@bartowski1182 yes, but not all models define that field, so I think a second guess based on the type of the first tensor type in the model will sometimes be necessary. Would you like some kind of `--outtype auto-f16` even if it's not the default?

Reply:
I'd be very interested in that yes, if they set it wrongly I would have probably messed up conversion anyways, but ideally I'd like to be mostly auto bf16 (since so much llama3 is flooding the market) with a tiny bit of intelligence to avoid it if it's explicitly set to something else

To confirm, we're able to quantize from BF16 right? It's not just meant for running at full precision?

Reply:
> but ideally I'd like to be mostly auto bf16

Yeah, my idea for `--outtype auto-f16` would use `f16` only if the first tensor of the model has dtype `torch.float16`, and use `bf16` otherwise (so that `bf16` is also chosen when converting from `f32` weights).
Is `auto-bf16` a better name? But it's a ~~bit~~ byte longer... :)
(I think that simply `auto` would be too ambiguous)

> To confirm, we're able to quantize from BF16 right?

Yes, I can confirm quantizing **from** `bf16` works.

Reply:
I've added `--outtype auto-f16` to guess which 16-bit float type to use depending on the type of the first loaded tensor (which seems like a good enough heuristic). `f16` is only used if that first tensor has type `torch.float16`, and otherwise, `bf16` is used.

Since most models are converted to then be quantized, I feel like this would be a good default, even if `bfloat16` inference is CPU-only, but I've kept `f16` as default for now.

Reply:
@compilade I'm using llama.cpp as a black box so I don't understand how it works. Can you explain why bf16 is only used for CPU? Will there be GPU support in the future?

Reply:
@compilade I think '--outtype auto' might be fine, since at its core that's what it's doing

only thing is, and maybe i can do it myself if it's too much tech debt, is it possible to figure out the auto-chosen version earlier and use it for naming the outfile? I also kind of want to add an --out-dir option, since I like putting the conversion in a new location and naming it -f32.gguf/-f16.gguf but won't be able to make it programmatic with this setup

Reply:
> Can you explain why bf16 is only used for CPU? Will there be GPU support in the future?

@htdung167 This PR is only about conversion, and the convert script always has been CPU-only.
Inference (text generation) with `bf16` was worked on by @jart in #6412. A relevant comment from there regarding future GPU support would be <https://github.com/ggerganov/llama.cpp/pull/6412#issuecomment-2028967868>.

> I think '--outtype auto' might be fine, since at its core that's what it's doing

@bartowski1182 I agree, I've thought about this more, and I'll change this to `auto` instead of `auto-f16`. I don't think there will be `auto-anything-else` anyway.

>  is it possible to figure out the auto-chosen version earlier and use it for naming the outfile?

yes, this is already possible by checking the logs. It would also be possible to do automatically, but not everyone has the same naming conventions, so maybe the right way to do this would be with a `.format()` pattern? For example, `--outfile llama-3-8b-instruct.{ftype}.gguf`, or `--outfile llama-3-8b-instruct.{outtype}.gguf` or `--outfile llama-3-8b-instruct.{}.gguf`. Not sure which to support (all?), but it should be clearer than `%s`. It would also be possible to allow using `{FTYPE}` or `{OUTTYPE}` for upper-cased type names.

I see you've used `fp16` and `fp32` in the past, but this will use `f16` and `f32`, respectively, for these type names.

Reply:
yeah i've made the transition to f16/f32, was an oversight from me naming them 'fp'

a format option would be amazing

Reply:
Should `auto` simply try to be as lossless as possible? Like, if the model is originally in `f32`, make the output `f32`? Or should it always select a 16-bit type?

Reply:
My vote would be on compressing even if originally it was f32, if the person covering wants f32 they'll specify, otherwise presumably they're always converting with the intention of quantizing where it won't matter

Reply:
"Do no harm" should be the default.